runtime/compiler/optimizer/TreeLowering.cpp

```diff
@@ -1461,3 +1432,57 @@ TR::TreeLowering::lowerStoreArrayElement(TR::PreorderNodeIterator& nodeIter, TR:
       gotoAfterHelperCallNode->addChildren(&deps, 1);
       }
    }
+
+/**
+ * @brief Perform lowering related to Valhalla value types
+ *
+ */
+void
+TR::TreeLowering::lowerValueTypeOperations(TransformationManager& transformations, TR::Node* node, TR::TreeTop* tt)
+   {
+   TR::SymbolReferenceTable * symRefTab = comp()->getSymRefTab();
+   static char *disableInliningCheckAastore = feGetEnv("TR_DisableVT_AASTORE_Inlining");
+
+   if (node->getOpCode().isCall())
+      {
+      if (symRefTab->isNonHelper(node->getSymbolReference(), TR::SymbolReferenceTable::objectEqualityComparisonSymbol))
+         {
+         // turn the non-helper call into a VM helper call
+         node->setSymbolReference(symRefTab->findOrCreateAcmpHelperSymbolRef());
+         static const bool disableAcmpFastPath =  NULL != feGetEnv("TR_DisableVT_AcmpFastpath");
+         if (!disableAcmpFastPath)
+            {
+            transformations.addTransformation(getTransformer<AcmpTransformer>(), node, tt);
+            }
+         }
+      else if (node->getSymbolReference()->getReferenceNumber() == TR_ldFlattenableArrayElement)
+         {
+         static char *disableInliningCheckAaload = feGetEnv("TR_DisableVT_AALOAD_Inlining");
+
+         if (!disableInliningCheckAaload)
+            {
+            TR_ASSERT_FATAL_WITH_NODE(tt->getNode(),
+                                      (tt->getNode()->getOpCodeValue() == TR::treetop) || (tt->getNode()->getOpCodeValue() == TR::NULLCHK),
+                                      "LoadArrayElementTransformer cannot process the treetop node that is neither a treetop nor a NULLCHK\n");
+
+            transformations.addTransformation(getTransformer<LoadArrayElementTransformer>(), node, tt);
+            }
+         }
+      else if (node->getSymbolReference()->getReferenceNumber() == TR_strFlattenableArrayElement)
+         {
+         if (!disableInliningCheckAastore)
+            {
+            TR_ASSERT_FATAL_WITH_NODE(tt->getNode(),
+                                      (tt->getNode()->getOpCodeValue() == TR::treetop) || (tt->getNode()->getOpCodeValue() == TR::NULLCHK),
+                                      "StoreArrayElementTransformer cannot process the treetop node that is neither a treetop nor a NULLCHK\n");
+            transformations.addTransformation(getTransformer<StoreArrayElementTransformer>(), node, tt);
+            }
+         }
+      }
+   // If inlining check for aastore is enabled, the NULLCHK on the value reference is
+   // taken care of by StoreArrayElementTransformer.
+   else if (node->getOpCodeValue() == TR::ArrayStoreCHK && disableInliningCheckAastore)
+      {
+      transformations.addTransformation(getTransformer<ArrayStoreCHKTransformer>(), node, tt);
+      }
```

Comment:
>It feels like the Transformer objects could be made responsible for performing these checks to see whether they can be applied, rather than having lowerValueTypeOperations encode that knowledge.

I was debating where or not to do it.  On one hand, I feel it's not necessary to add the transformations if it's already disabled. On another hand, the logic in `lowerValueTypeOperations` is getting a bit bulky. I agree this is something to consider when it's generalized to VP or if more tree lowering Transformers are added.